hack/test-go.sh

```diff
@@ -1,125 +1,231 @@
 
-  rm -rf $OUTPUT_COVERAGE/$OS_GO_PACKAGE
-else
-  nice go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "${@:2}" $test_packages
+    test_packages="$(echo "${openshift_test_packages}"; echo "${mandatory_kubernetes_packages}")"
+
+    if [[ -n "${test_kube}" ]]; then
+        optional_kubernetes_packages="$(find "${kubernetes_path}" -not \( \
+          \( \
+            -wholename "${kubernetes_path}/pkg/api" \
+            -o -wholename "${kubernetes_path}/pkg/api/v1" \
+            -o -wholename "${kubernetes_path}/test/e2e" \
+          \) -prune \
+        \) -name '*_test.go' -print0 | xargs -0n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n")"
+
+        test_packages="$(echo "${test_packages}"; echo "${optional_kubernetes_packages}")"
+    fi
 fi
 
-ret=$?; ENDTIME=$(date +%s); echo "$0 took $(($ENDTIME - $STARTTIME)) seconds"; exit "$ret"
+if [[ -n "${print_packages}" ]]; then
+    echo "The following packages will be tested by $0:"
+    for package in "${test_packages}"; do
+        echo "${package}"
+    done
+    exit 0
+fi
+
+# Run 'go test' with the accumulated arguments and packages:
+if [[ -n "${junit_report}" ]]; then
+    # we need to generate jUnit xml
+    hack/build-go.sh tools/junitreport
+    junitreport="$(os::build::find-binary junitreport)"
+
+    if [[ -z "${junitreport}" ]]; then
+        echo "It looks as if you don't have a compiled junitreport binary"
+        echo
+        echo "If you are running from a clone of the git repo, please run"
+        echo "'./hack/build-go.sh tools/junitreport'."
+        exit 1
+    fi
+
+    test_output_file="${LOG_DIR}/test-go.log"
+    junit_report_file="${ARTIFACT_DIR}/report.xml"
+
+    # we don't care if 'go test' fails, we want to summarize the output anyway.
+    set +o errexit
+    go test ${gotest_flags} ${test_packages} > ${test_output_file} 2>&1
+    return_code=$?
+    set -o errexit
+
+    cat "${test_output_file}" | "${junitreport}" --type gotest --suites nested --roots github.com/openshift/origin > "${junit_report_file}"
+    cat "${junit_report_file}" | "${junitreport}" summarize
+    exit "${return_code}"
+
+elif [[ -n "${coverage_output_dir}" ]]; then
+    # we need to generate coverage reports
+    for test_package in "${test_packages}"; do
+        mkdir -p "${coverage_output_dir}/${test_package}"
+        gotest_flags="${gotest_flags} -coverprofile=${coverage_output_dir}/${test_package}/profile.out"
+
+        go test ${gotest_flags} ${test_package}
+
+        # assemble all profiles and generate a coverage report
+        echo 'mode: atomic' > "${coverage_output_dir}/profiles.out"
+        find "${coverage_output_dir}" -name profile.out | xargs sed '/^mode: atomic$/d' >> "${coverage_output_dir}/profiles.out"
+
+        go tool cover "-html=${coverage_output_dir}/profiles.out" -o "${coverage_output_dir}/coverage.html"
+
+        rm -rf "${coverage_output_dir}/${OS_GO_PACKAGE}"
```

Comment:
Is this supposed to be `"${coverage_output_dir}/${test_package}"` to clean up the directory created at the start of the `for` block, or is `${coverage_output_dir}/${OS_GO_PACKAGE}` something that `go tool cover` creates? If the latter, it might be worth adding a comment to that effect.